cloud/blockstore/libs/storage/disk_agent/disk_agent_state.cpp

```diff
@@ -815,4 +820,84 @@ void TDiskAgentState::StopTarget()
     }
 }
 
+void TDiskAgentState::UpdateSessionCache(TDeviceClient& client) const
+{
+    const auto path = AgentConfig->GetCachedSessionsPath();
+
+    if (path.empty()) {
+        STORAGE_INFO("Session cache is not configured.");
+        return;
+    }
+
+    try {
+        auto sessions = client.GetSessions();
+
+        NProto::TDiskAgentDeviceSessionCache proto;
+        proto.MutableSessions()->Assign(
+            std::make_move_iterator(sessions.begin()),
+            std::make_move_iterator(sessions.end())
+        );
+
+        const TString tmpPath {path + ".tmp"};
+
+        SerializeToTextFormat(proto, tmpPath);
+
+        if (!NFs::Rename(tmpPath, path)) {
+            const auto ec = errno;
+            ythrow TServiceError {MAKE_SYSTEM_ERROR(ec)} << strerror(ec);
+        }
+    } catch (...) {
+        STORAGE_ERROR("Can't update session cache: " << CurrentExceptionMessage());
+        ReportDiskAgentSessionCacheUpdateError();
+    }
+}
+
+void TDiskAgentState::RestoreSessions(TDeviceClient& client) const
+{
+    const auto path = AgentConfig->GetCachedSessionsPath();
+
+    if (path.empty()) {
+        STORAGE_INFO("Session cache is not configured.");
+        return;
+    }
+
+    if (!NFs::Exists(path)) {
+        STORAGE_INFO("Session cache is empty.");
+        return;
+    }
+
+    NProto::TDiskAgentDeviceSessionCache proto;
+
+    try {
+        ParseProtoTextFromFileRobust(path, proto);
+
+        auto& sessions = *proto.MutableSessions();
+
+        STORAGE_INFO("Found " << sessions.size()
+            << " sessions in the session cache:\n" << JoinSeq("\n", sessions));
+
+        for (auto& session: sessions) {
+            TVector<TString> uuids(
+                std::make_move_iterator(session.MutableDeviceIds()->begin()),
+                std::make_move_iterator(session.MutableDeviceIds()->end()));
+
+            CheckError(client.AcquireDevices(
```

Comment:
Before calling to `client.AcquireDevices` we need to filter out uuids from old devices that were removed from the disk agent.
Otherwise disk agent can be stopped and some change in config will remove device. Then we will try to acquire the old device from the saved sessions file and it will fail because of the code here https://github.com/ydb-platform/nbs/blob/main/cloud/blockstore/libs/storage/disk_agent/model/device_client.cpp#L40

Also if we won't prune old devices that stopped to be part of the disk agent config we might end with large sessions file containing sessions with invalid device uuids

Reply:
I've tweaked the code so that only broken sessions will be rejected now. There is no problem with old devices because `UpdateSessionCache` rewrites the cache file completely with actual sessions.